File: LeetCode/451-sort-characters-by-frequency.py
###############################################################################################
# easy
###########
# 时间复杂度：O(n)
# 空间复杂度：O(n)
###############################################################################################
class Solution:
    def frequencySort(self, s: str) -> str:
        dic = defaultdict(int)
        for ch in s:
            dic[ch] += 1
        lis = []
        for key in dic:
            lis.append((key, dic[key]))
        lis.sort(key=lambda x: -x[1])
        return "".join([per[0]*per[1] for per in lis])

File: LeetCode/451-sort-characters-by-frequency.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// hard thing is to write the sort in c++ syntax
////////////////
// Time Complexity: O(nlogn)
// Space Complexity: O(logn)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    string frequencySort(string s) {
        int counts[256] = {0};
        // record frequency
        for (char ch : s) counts[ch] ++;
        
        sort(s.begin(), s.end(), [&](char a, char b) { 
            return counts[a] > counts[b] || (counts[a] == counts[b] && a < b); 
        });
        return s;
    }
};